pkg/executables/cmk.go

```diff
@@ -119,39 +135,147 @@ func (c *Cmk) ValidateAffinityGroupsPresent(ctx context.Context, domain string,
+		if err != nil {
+			return nil, fmt.Errorf("error getting zones info: %v", err)
+		}
+		if result.Len() == 0 {
+			return nil, fmt.Errorf("zone %s not found", filterString)
+		}
+
+		response := struct {
+			CmkZones []cmkZone `json:"zone"`
+		}{}
+		if err = json.Unmarshal(result.Bytes(), &response); err != nil {
+			return nil, fmt.Errorf("failed to parse response into json: %v", err)
+		}
+		cmkZones := response.CmkZones
+		if len(cmkZones) > 1 {
+			return nil, fmt.Errorf("duplicate zone %s found", filterString)
+		} else if len(zones) == 0 {
+			return nil, fmt.Errorf("zone %s not found", filterString)
+		} else {
+			zoneIdentifiers = append(zoneIdentifiers, v1alpha1.CloudStackResourceIdentifier{Name: cmkZones[0].Name, Id: cmkZones[0].Id})
+		}
+	}
+	return zoneIdentifiers, nil
+}
+
+func (c *Cmk) ValidateDomainPresent(ctx context.Context, domain string) (v1alpha1.CloudStackResourceIdentifier, error) {
+	domainIdentifier := v1alpha1.CloudStackResourceIdentifier{Name: domain, Id: ""}
+	result, err := c.exec(ctx, "list", "domains", fmt.Sprintf("name=\"%s\"", domain))
+	if err != nil {
+		return domainIdentifier, fmt.Errorf("error getting domain info: %v", err)
+	}
+	if result.Len() == 0 {
+		return domainIdentifier, fmt.Errorf("domain %s not found", domain)
+	}
+
+	response := struct {
+		CmkDomains []cmkDomain `json:"domain"`
+	}{}
+	if err = json.Unmarshal(result.Bytes(), &response); err != nil {
+		return domainIdentifier, fmt.Errorf("failed to parse response into json: %v", err)
+	}
+	domains := response.CmkDomains
+	if len(domains) > 1 {
+		return domainIdentifier, fmt.Errorf("duplicate domain %s found", domain)
+	} else if len(domains) == 0 {
+		return domainIdentifier, fmt.Errorf("domain %s not found", domain)
+	}
+
+	domainIdentifier.Id = domains[0].Id
+	domainIdentifier.Name = domains[0].Name
+
+	return domainIdentifier, nil
+}
+
+func (c *Cmk) ValidateNetworkPresent(ctx context.Context, domainId string, zoneRef v1alpha1.CloudStackZoneRef, zones []v1alpha1.CloudStackResourceIdentifier, account string, multipleZone bool) error {
+	filterArgs := []string{"list", "networks"}
+	if zoneRef.Network.Type == v1alpha1.Id {
+		filterArgs = append(filterArgs, fmt.Sprintf("id=\"%s\"", zoneRef.Network.Value))
+	}
+	if multipleZone {
+		filterArgs = append(filterArgs, fmt.Sprintf("type=\"%s\"", "Shared"))
+	}
+	// account must be specified within a domainId
+	// domainId can be specified without account
+	if len(domainId) > 0 {
+		filterArgs = append(filterArgs, fmt.Sprintf("domainid=\"%s\"", domainId))
+		if len(account) > 0 {
+			filterArgs = append(filterArgs, fmt.Sprintf("account=\"%s\"", account))
+		}
+	}
+	zoneId := ""
+	if zoneRef.Zone.Type == v1alpha1.Id {
+		zoneId = zoneRef.Zone.Value
 	} else {
-		filterArg = fmt.Sprintf("name=\"%s\"", zone.Value)
+		for _, zoneIdentifier := range zones {
+			if zoneRef.Zone.Value == zoneIdentifier.Name {
+				zoneId = zoneIdentifier.Id
+				break
+			}
+		}
```

Comment:
it seems like the retrieval of zoneID should be extracted into a separate function in case we need to do it elsewhere